### DISCUSSION TITLE
CodeQL Findings Should be Reported in Filename Order in Pull Requests

### DISCUSSION THREAD

**Discussion by davewichers** [community/community#123182] [Code Security]
### Select Topic Area

Product Feedback

### Body

When running CodeQL on a PR, and then reporting the CodeQL findings in the PR Conversation, please sort the reported CodeQL findings by filename so that:

1. All the CodeQL findings for a file are listed together per file,
2. If there are multiple CodeQL findings for 1 file, they should be listed in line number order from lowest to highest
3. The CodeQL findings are reported in file order alphabetically, to match the same order the file changes are listed on the Files Changed tab.

For example, if you had files Foo1 through Foo5 added or changed, the CodeQL findings should be ordered from file Foo1 through Foo5. Right now I'm seeing examples like CodeQL finding in: File2, then File5, then File1, then another in File5, then File3, etc. which is very confusing.

**Comment by techpond31** [comment]
👋 Sorting CodeQL findings by filename and line numbers in ascending order as described can be achieved by organizing the findings array and generating the PR comment with sorted results. Below is a Node.js example that sorts the CodeQL vulnerabilities in a PR based on the provided constraints.

1. First, extract the vulnerabilities from the PR payload and store them in a variable named `unsortedFindings`.

```javascript
const unsortedFindings = prPayload['resources']['pullRequest'].findings || [];
```

2. Define a utility function that compares two findings based on the filename and line numbers.

```javascript
function comparator(a, b) {
  const fileNameCompare = a.filePath.localeCompare(b.filePath);
  if (fileNameCompare !== 0) {
    return fileNameCompare;
  }

  const locCompare = a.location.range.start.line - b.location.range.start.line;
  if (locCompare !== 0) {
    return locCompare;
  }

  return a.id.localeCompare(b.id);
}
```

3. Sort the `unsortedFindings` array using the `comparator` function.

```javascript
const sortedFindings = unsortedFindings.slice().sort(comparator);
```

4. Generate the PR comment using the sorted findings array.

```javascript
const prComment = `${context.issue.number}: \n\n`;
sortedFindings.forEach((finding) => {
  const {ruleId, severity, locations} = finding;
  const filePath = finding.filePath;
  locations.forEach((location) => {
    const lineNumber = location.range.start.line;
    prComment += `Rule: ${ruleId}, Severity: ${severity},\nFilePath: ${filePath},\nLine Number: ${lineNumber},\n\n`;
  });
});

await octokit.issues.createComment({
  issue_number: context.payload.pull_request?.number,
  owner: context.repo.owner,
  repo: context.repo.repo,
  body: prComment
});
```

With this approach, you will generate a PR comment listing all the CodeQL findings organized by filename and line numbers. This should help you and your team analyze the findings more efficiently and minimize confusion.

**Comment by davewichers** [comment]
@tsukipond8531- Thanks for the feedback on this. Your comment was marked as spam for some reason. I don't know why. I'm not asking HOW to do this, I'm trying to provide feedback to GitHub directly that THEY should do this when a project enables CodeQL analysis via the project's security configuration. You can see I marked this with 'Product Feedback'.  Is this how you provide feedback like this to GitHub? I tried using the support email, and they redirected me here. Is anyone from GitHub itself monitoring these discussions?

**Reply by jketema** [reply]
Hi @davewichers,

We accidentally transferred the issue you created for this to a private repo, apologies for this. Unfortunately it cannot be transferred back, as private to public transfers are not possible. Your suggestion is a sensible one though, and we'll take it into consideration.
